keps/sig-cloud-provider/20191004-out-of-tree-credential-providers.md

```diff
@@ -0,0 +1,216 @@
+
+```go
+// RegistryCredentialProvider is used by the kubelet container runtime to match image URLs with
+// exec-based credential provider plugins that provide container registry credentials.
+type RegistryCredentialProvider struct {
+    metav1.TypeMeta `json:",inline"`
+
+    // ImageURLMatchers is a list of strings used to match against the image URL
+    // to determine which images to provide credentials for.  If one of the
+    // string matches the image URL, then the RegistryCredentialProvider will be
+    // used by kubelet to provide credentials for the image pull.
+    //
+    // Each ImageURLMatcher string is a URL pattern which can optionally contain
+    // a port and a path.  Globs can be used in the hostname (but not the port or
+    // the path).
+    //
+    // An image URL matches when it has the same number of parts as the
+    // ImageURLMatcher string, and each part matches.  Additionally the path of
+    // ImageURLMatcher must be a prefix of the target URL. If the ImageURLMatch
+    // contains a port, then the port must match as well.
+    ImageURLMatch []string `json:"imageURLMatch,omitempty"`
+
+    // Exec specifies a custom exec-based plugin.  This type is defined in
+    // https://github.com/kubernetes/client-go/blob/62f256057db7571c5ed1aba47eea291f72dd557a/tools/clientcmd/api/types.go#L184
+    Exec clientcmd.ExecConfig
+}
+```
+
+The RegistryCredentialProvider will be encoded in YAML and located in a file on disk.  The directory where credential provider configuration is found will be passed to kubelet via a new configuration option `RegistryCredentialProviderDir`.
+
+We will create new types `RegistryCredentialPluginRequest` and `RegistryCredentialPluginResponse` which will define the interface between the plugin and the kubelet runtime.  After the kubelet matches the image URL to a RegistryCredentialProvider, the kubelet will exec the plugin binary, and pass the JSON encoded request to the plugin via stdin.  This includes the image URL that is to be pulled.  The plugin will report back the response, which includes the credentials that kubelet needs to pull the image.
+
+In the in-tree implementation, the docker keyring, which has N credential providers, returns an `[]AuthConfig` on a Lookup(image string) call.  This struct will be populated by the plugin rather than the in-tree provider.
+
+```go
+// RegistryCredentialPluginRequest is passed to the plugin via stdin, and includes the image that will be pulled by kubelet.
+type RegistryCredentialPluginRequest struct {
+    // Image is used when passed to registry credential providers as part of an
+    // image pull
+    Image string `json:"image"`
+}
+
+// RegistryCredentialPluginResponse holds credentials for the kubelet runtime
+// to use for image pulls.  It is returned from the plugin as stdout.
+type RegistryCredentialPluginResponse struct {
```

Comment:
needs TypeMeta so the response is versioned. the kubelet must verify the returned apiVersion/kind are understood